src/assembly/collections.ts

```diff
@@ -0,0 +1,157 @@
+import * as utils from "./utils";
+
+export class CollectionMutationResult {
+  collection!: string;
+  operation!: string;
+  status!: string;
+  key!: string;
+  error!: string;
+}
+
+export class SearchMethodMutationResult {
+  collection!: string;
+  searchMethod!: string;
+  operation!: string;
+  status!: string;
+  error!: string;
+}
+
+export class CollectionSearchResult {
+  collection!: string;
+  searchMethod!: string;
+  status!: string;
+  objects!: CollectionSearchResultObject[];
+  error!: string;
+}
+
+export class CollectionSearchResultObject {
+  key!: string;
+  text!: string;
+  score!: f64;
+}
+
+// @ts-expect-error: decorator
+@external("hypermode", "upsertToCollection")
+declare function hostUpsertToCollection(
+  collection: string,
+  key: string | null,
+  text: string,
+): CollectionMutationResult;
+
+// @ts-expect-error: decorator
+@external("hypermode", "deleteFromCollection")
+declare function hostDeleteFromCollection(
+  collection: string,
+  key: string,
+): CollectionMutationResult;
+
+// @ts-expect-error: decorator
+@external("hypermode", "searchCollection")
+declare function hostSearchCollection(
+  collection: string,
+  searchMethod: string,
+  text: string,
+  limit: i32,
+  returnText: bool,
+): CollectionSearchResult;
+
+// @ts-expect-error: decorator
+@external("hypermode", "recomputeSearchMethod")
+declare function hostRecomputeSearchMethod(
+  collection: string,
+  searchMethod: string,
+): SearchMethodMutationResult;
+
+// @ts-expect-error: decorator
+@external("hypermode", "computeSimilarity")
+declare function hostComputeSimilarity(
+  collection: string,
+  searchMethod: string,
+  key1: string,
+  key2: string,
+): CollectionSearchResultObject;
+
+// @ts-expect-error: decorator
+@external("hypermode", "getText")
+declare function hostGetText(collection: string, key: string): string;
+
+// @ts-expect-error: decorator
+@external("hypermode", "getTexts")
+declare function hostGetTexts(collection: string): Map<string, string>;
+
+// add data to in-mem storage, get all embedders for a collection, run text through it
+// and insert the Text into the Text indexes for each search method
```

Comment:
FYI - If you were intending these comments to be visible for typeahead/hover in VS Code, then you need to use JSDoc style comment blocks.

```ts
/**
 * example
 */
```

Tags for the parameters and return types are also supported.   We can always do this later though.